src/models/Satisfiability.jl

```diff
@@ -0,0 +1,198 @@
+
+julia> satisfiable(cnf, Dict([:x=>false, :y=>false, :z=>true]))
+false
+```
+"""
+struct CNF{T}
+    clauses::Vector{CNFClause{T}}
+end
+function Base.show(io::IO, c::CNF)
+    print(io, join(["($k)" for k in c.clauses], " ∧ "))
+end
+Base.:(==)(x::CNF, y::CNF) = x.clauses == y.clauses
+Base.length(x::CNF) = length(x.clauses)
+
+"""
+    ¬(var::BoolVar)
+
+Negation of a boolean variables of type [`BoolVar`](@ref).
+"""
+¬(var::BoolVar{T}) where T = BoolVar(var.name, ~var.neg)
+
+"""
+    ∨(vars...)
+
+Logical `or` applied on [`BoolVar`](@ref) and [`CNFClause`](@ref).
+Returns a [`CNFClause`](@ref).
+"""
+∨(var::BoolVar{T}, vars::BoolVar{T}...) where T = CNFClause([var, vars...])
+∨(c::CNFClause{T}, var::BoolVar{T}) where T = CNFClause([c.vars..., var])
+∨(c::CNFClause{T}, d::CNFClause{T}) where T = CNFClause([c.vars..., d.vars...])
+∨(var::BoolVar{T}, c::CNFClause) where T = CNFClause([var, c.vars...])
+
+
+"""
+    ∧(vars...)
+
+Logical `and` applied on [`CNFClause`](@ref) and [`CNF`](@ref).
+Returns a new [`CNF`](@ref).
+"""
+∧(c::CNFClause{T}, cs::CNFClause{T}...) where T = CNF([c, cs...])
+∧(c::CNFClause{T}, cs::CNF{T}) where T = CNF([c, cs.clauses...])
+∧(cs::CNF{T}, c::CNFClause{T}) where T = CNF([cs.clauses..., c])
+∧(cs::CNF{T}, ds::CNF{T}) where T = CNF([cs.clauses..., ds.clauses...])
+
+"""
+    @bools(syms::Symbol...)
+
+Create some boolean variables of type [`BoolVar`](@ref) in current scope that can be used in create a [`CNF`](@ref).
+
+Example
+------------------------
+```jldoctest; setup=:(using GenericTensorNetworks)
+julia> @bools x y z
+
+julia> (x ∨ y) ∧ (¬y ∨ z)
+(x ∨ y) ∧ (¬y ∨ z)
+```
+"""
+macro bools(syms::Symbol...)
+    esc(Expr(:block, [:($s = $BoolVar($(QuoteNode(s)))) for s in syms]..., nothing))
+end
+
+
+"""
+The formal definition of SAT problem
+
+The [satisfiability](https://queracomputing.github.io/GenericTensorNetworks.jl/dev/generated/Satisfiability/) problem.
+
+Positional arguments
+-------------------------------
+* `cnf` is a conjunctive normal form ([`CNF`](@ref)) for specifying the satisfiability problems.
+* `weights` are associated with clauses.
+
+Examples
+-------------------------------
+```jldoctest; setup=:(using GenericTensorNetworks)
+julia> @bools x y z a b c
+
+julia> c1 = x ∨ ¬y
+x ∨ ¬y
+
+julia> c2 = c ∨ (¬a ∨ b)
+c ∨ ¬a ∨ b
+
+julia> c3 = (z ∨ ¬a) ∨ y
+z ∨ ¬a ∨ y
+
+julia> c4 = (c ∨ z) ∨ ¬b
+c ∨ z ∨ ¬b
+
+julia> cnf = (c1 ∧ c4) ∧ (c2 ∧ c3)
+(x ∨ ¬y) ∧ (c ∨ z ∨ ¬b) ∧ (c ∨ ¬a ∨ b) ∧ (z ∨ ¬a ∨ y)
+
+```
+"""
+struct Satisfiability{T} <:AbstractProblem
+    cnf::CNF{T}
+    function Satisfiability(cnf::CNF{T}) where {T}
+        new{T}(cnf)
+    end
+end
+
+function variables(c::Satisfiability)
+    var_names = Set{}()
+    for clause in c.cnf.clauses
+        for var in clause.vars
+            push!(var_names, var.name)
+        end
+    end
+    return collect(var_names) 
+end
+flavors(::Type{<:Satisfiability}) = [0, 1]  # false, true
+terms(c::Satisfiability) = collect(c.cnf.clauses)
```

Comment:
NOTE: we may consider removing the `terms` interface. This is only required for energy based models.

Reply:
Please check #38